spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestSparkTruncateFunction.java

```diff
@@ -0,0 +1,436 @@
+    Assert.assertEquals(-2L, scalarSql("SELECT system.truncate(2, -1L)"));
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS bigint))"));
+  }
+
+  @Test
+  public void testTruncateDecimal() {
+    // decimal truncation works by applying the decimal scale to the width: ie 10 scale 2 = 0.10
+    Assert.assertEquals(
+        new BigDecimal("12.30"),
+        scalarSql("SELECT system.truncate(10, CAST(%s as DECIMAL(9, 2)))", "12.34"));
+
+    Assert.assertEquals(
+        new BigDecimal("12.30"),
+        scalarSql("SELECT system.truncate(10, CAST(%s as DECIMAL(9, 2)))", "12.30"));
+
+    Assert.assertEquals(
+        new BigDecimal("12.290"),
+        scalarSql("SELECT system.truncate(10, CAST(%s as DECIMAL(9, 3)))", "12.299"));
+
+    Assert.assertEquals(
+        new BigDecimal("0.03"),
+        scalarSql("SELECT system.truncate(3, CAST(%s as DECIMAL(5, 2)))", "0.05"));
+
+    Assert.assertEquals(
+        new BigDecimal("0.00"),
+        scalarSql("SELECT system.truncate(10, CAST(%s as DECIMAL(9, 2)))", "0.05"));
+
+    Assert.assertEquals(
+        new BigDecimal("-0.10"),
+        scalarSql("SELECT system.truncate(10, CAST(%s as DECIMAL(9, 2)))", "-0.05"));
+
+    Assert.assertEquals(
+        "Implicit decimal scale and precision should be allowed",
+        new BigDecimal("12345.3480"),
+        scalarSql("SELECT system.truncate(10, 12345.3482)"));
+
+    BigDecimal truncatedDecimal =
+        (BigDecimal) scalarSql("SELECT system.truncate(10, CAST(%s as DECIMAL(6, 4)))", "-0.05");
+    Assert.assertEquals(
+        "Truncating a decimal should return a decimal with the same scale",
+        4,
+        truncatedDecimal.scale());
+
+    Assert.assertEquals(
+        "Truncating a decimal should return a decimal with the correct scale",
+        BigDecimal.valueOf(-500, 4),
+        truncatedDecimal);
+
+    Assert.assertEquals(
+        "Null input should return null",
+        null,
+        scalarSql("SELECT system.truncate(2, CAST(null AS decimal))"));
+  }
+
+  @Test
+  public void testTruncateString() {
+    Assert.assertEquals(
+        "Should system.truncate strings longer than length",
+        "abcde",
+        scalarSql("SELECT system.truncate(5, 'abcdefg')"));
+
+    Assert.assertEquals(
+        "Should not pad strings shorter than length",
+        "abc",
+        scalarSql("SELECT system.truncate(5, 'abc')"));
+
+    Assert.assertEquals(
+        "Should not alter strings equal to length",
+        "abcde",
+        scalarSql("SELECT system.truncate(5, 'abcde')"));
+
+    Assert.assertEquals(
+        "Should handle three-byte UTF-8 characters appropriately",
+        "测",
+        scalarSql("SELECT system.truncate(1, '测试')"));
+
+    Assert.assertEquals(
+        "Should handle three-byte UTF-8 characters mixed with two byte utf-8 characters",
+        "测试ra",
+        scalarSql("SELECT system.truncate(4, '测试raul试测')"));
```

Comment:
Because we are using Spark's truncate function, I think this needs to run through all of the test cases for Iceberg's `UnicodeUtil.truncateString` and make sure we get the expected output.

Reply:
I agree with that.

But I took a look and I actually couldn't find any test cases for `UnicodeUtil.truncateString` directly.

I looked up the tests for `TestTruncate`, but `TestTruncate#testTruncateString` has _very few_ tests - all of which are covered here.

https://github.com/apache/iceberg/blob/5a15efc070ab59eeda6343998aa065c0c9892c5c/api/src/test/java/org/apache/iceberg/transforms/TestTruncate.java#L70-L77

Reply:
Let me see if I can find more test cases somewhere 🙂

Reply:
`TestMetricsTruncation` has more test cases, though they're specifically for `testTruncateMin` and `testTruncateMax`.

But I can reuse those as they do cover more situations: https://github.com/apache/iceberg/blob/5a15efc070ab59eeda6343998aa065c0c9892c5c/core/src/test/java/org/apache/iceberg/TestMetricsTruncation.java#L132-L264

Reply:
I copied over the tests from `TestmetricsTruncation` to ensure that 2-byte, 3-byte, and 4-byt UTF-8 codepoints were all covered.

So we should be fully covered in terms of testing truncate String (at least as covered as we are elsewhere).

Reply:
Looks good.